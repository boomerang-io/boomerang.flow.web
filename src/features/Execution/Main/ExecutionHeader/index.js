import React from "react";
import PropTypes from "prop-types";
import { withRouter, Link } from "react-router-dom";
import { SkeletonPlaceholder } from "carbon-components-react";
import FeatureHeader from "Components/FeatureHeader";
import moment from "moment";
import { REQUEST_STATUSES } from "Config/servicesConfig";
import styles from "./executionHeader.module.scss";

ExecutionHeader.propTypes = {
  workflow: PropTypes.object.isRequired,
  workflowExecution: PropTypes.object.isRequired
};

function ExecutionHeader({ history, workflow, workflowExecution }) {
  const { state } = history.location;
  const { teamName, initiatedByUserName, trigger, creationDate } = workflowExecution.data;

  return (
    <FeatureHeader includeBorder>
      <div className={styles.container}>
        <section>
          <div className={styles.subtitle}>
            <Link to={state ? state.fromUrl : "/activity"}>{state ? state.fromText : "Activity"}</Link>
            <p style={{ margin: "0 0.5rem" }}>/</p>
            {workflow.isFetching ? (
              <SkeletonPlaceholder className={styles.workflowNameSkeleton} />
            ) : (
              <p>{workflow.data.name}</p>
            )}
          </div>
          <h1 className={styles.title}>Workflow run detail</h1>
        </section>
        {workflowExecution.status === REQUEST_STATUSES.SUCCESS ? (
          <div className={styles.content}>
            <div className={styles.data}>
              <p className={styles.dataTitle}>Team</p>
              <p className={styles.dataValue}>{teamName}</p>
            </div>
            <div className={styles.data}>
              <p className={styles.dataTitle}>Initiated by</p>
              {initiatedByUserName ? (
                <p className={styles.dataValue}>{initiatedByUserName}</p>
              ) : (
                <span aria-label="robot" aria-hidden={false} role="img">
                  {"ðŸ¤–"}
                </span>
              )}
            </div>
            <div className={styles.data}>
              <p className={styles.dataTitle}>Trigger</p>
              <p className={styles.dataValue}>{trigger}</p>
            </div>
            <div className={styles.data}>
              <p className={styles.dataTitle}>Team</p>
              <p className={styles.dataValue}>{moment(creationDate).format("YYYY-MM-DD hh:mm A")}</p>
            </div>
          </div>
        ) : (
          <SkeletonPlaceholder className={styles.headerContentSkeleton} />
        )}
      </div>
    </FeatureHeader>
  );
}

export default withRouter(ExecutionHeader);
