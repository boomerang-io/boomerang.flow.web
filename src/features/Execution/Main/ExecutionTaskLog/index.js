import React, { useState } from "react";
import PropTypes from "prop-types";
import orderBy from "lodash/orderBy";
import { Arrows32, ChevronLeft32 } from "@carbon/icons-react";
import { SkeletonPlaceholder, TooltipIcon } from "carbon-components-react";
import TaskItem from "./TaskItem";
import { REQUEST_STATUSES } from "Config/servicesConfig";
import { ACTIVITY_STATUSES_TO_ICON, ACTIVITY_STATUSES_TO_TEXT } from "Constants/activityStatuses";
import { getSimplifiedDuration } from "Utilities/timeHelper";
import styles from "./executionTaskLog.module.scss";

ExecutionTaskLog.propTypes = {
  workflowExecution: PropTypes.object.isRequired
};

function ExecutionTaskLog({ workflowExecution }) {
  const [isCollapsed, setIsCollapsed] = useState(false);
  const [tasksSort, setTasksSort] = useState("desc");

  const { id, duration, status, steps } = workflowExecution.data;
  const Icon = ACTIVITY_STATUSES_TO_ICON[status];

  const toggleCollapse = () => {
    setIsCollapsed(!isCollapsed);
  };

  const toggleSort = () => {
    setTasksSort(tasksSort === "desc" ? "asc" : "desc");
  };

  const sortedTasks = steps ? orderBy(steps, step => step.order, [tasksSort]) : [];

  return (
    <aside className={`${styles.container} ${isCollapsed ? styles.collapsed : ""}`}>
      {workflowExecution.status === REQUEST_STATUSES.SUCCESS ? (
        <section className={`${styles.statusBlock} ${styles[status]}`}>
          <div className={styles.duration}>
            <p className={styles.title}>Duration</p>
            <time className={styles.value}>
              {typeof duration === "number" ? getSimplifiedDuration(duration / 1000) : "--"}
            </time>
          </div>
          <div className={styles.status}>
            <p className={styles.title}>Status</p>
            <div className={styles.statusData}>
              {Icon && <Icon aria-label={status} className={styles.statusIcon} />}
              <p className={styles.value}>{status ? ACTIVITY_STATUSES_TO_TEXT[status] : "--"}</p>
            </div>
          </div>
          <button className={styles.collapseButton} onClick={toggleCollapse}>
            <ChevronLeft32 className={styles.chevron} />
          </button>
        </section>
      ) : (
        <section className={styles.statusBlock}>
          <SkeletonPlaceholder className={styles.statusBlockSkeleton} />
        </section>
      )}
      <section className={styles.taskbar}>
        <p className={styles.taskbarTitle}>Task log</p>
        {!isCollapsed && (
          <TooltipIcon
            align="center"
            direction="top"
            tooltipText="Change sort direction (by start time)"
          >
            <button onClick={toggleSort} className={styles.taskbarButton} data-testid="taskbar-button">
              <Arrows32 className={styles.taskbarArrows} />
            </button>
          </TooltipIcon>
        )}
      </section>
      <ul className={styles.tasklog}>
        {workflowExecution.status === REQUEST_STATUSES.SUCCESS ? (
          sortedTasks.map(step => <TaskItem key={step.id} flowActivityId={id} hidden={isCollapsed} task={step} />)
        ) : (
          <SkeletonPlaceholder className={styles.taskLogSkeleton} />
        )}
      </ul>
    </aside>
  );
}

export default ExecutionTaskLog;
