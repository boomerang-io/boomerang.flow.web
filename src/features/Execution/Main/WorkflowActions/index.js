import React from "react";
import PropTypes from "prop-types";
import { useHistory } from "react-router-dom";
import { Edit32 } from "@carbon/icons-react";
import styles from "./WorkflowActions.module.scss";

WorkflowActions.propTypes = {
  workflow: PropTypes.object.isRequired
};

function WorkflowActions({ workflow }) {
  let history = useHistory();
  const { id } = workflow;

  return (
    <div className={styles.container}>
      <p className={styles.messageText}>Read-only</p>
      <button className={styles.editContainer} onClick={() => history.push(`/editor/${id}/designer`)}>
        <Edit32 className={styles.editIcon} />
        <p className={styles.editText}>Edit Workflow</p>
      </button>
    </div>
  );
}

export default WorkflowActions;
