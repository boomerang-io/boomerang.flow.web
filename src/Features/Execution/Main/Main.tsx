import React, { useEffect, useRef } from "react";
import { DiagramWidget } from "@projectstorm/react-diagrams";
import { Loading } from "@boomerang-io/carbon-addons-boomerang-react";
import ExecutionHeader from "./ExecutionHeader";
import ExecutionTaskLog from "./ExecutionTaskLog";
import WorkflowActions from "./WorkflowActions";
import WorkflowZoom from "Components/WorkflowZoom";
import WorkflowDagEngine from "Utils/dag/WorkflowDagEngine";
import { ExecutionStatus, QueryStatus, WorkflowDagEngineMode } from "Constants";
import { WorkflowSummary, WorkflowRevision, WorkflowDetailedActivityExecution } from "Types";
import styles from "./main.module.scss";

interface MainProps {
  dag: WorkflowRevision["dag"];
  workflow: {
    data: WorkflowSummary;
    status: string;
    error: any;
  };
  workflowExecution: {
    data: WorkflowDetailedActivityExecution;
    status: string;
    error: any;
  };
}

const Main: React.FC<MainProps> = ({ dag, workflow, workflowExecution }) => {
  const workflowDagEngine = new WorkflowDagEngine({
    dag,
    mode: WorkflowDagEngineMode.Executor,
  });
  const diagramRef = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    workflowDagEngine.getDiagramEngine().zoomToFit();
  }, [workflowDagEngine]);

  const { status, steps } = workflowExecution.data;

  const hasFinished = [ExecutionStatus.Completed, ExecutionStatus.Invalid, ExecutionStatus.Failure].includes(status);

  const hasStarted = steps && steps.find((step) => step.flowTaskStatus !== ExecutionStatus.NotStarted);

  const isDiagramLoading =
    workflow.status === QueryStatus.Success &&
    workflowExecution.status === QueryStatus.Success &&
    (hasStarted || hasFinished);

  const workflowDagBoundingClientRect = diagramRef.current ? diagramRef.current.getBoundingClientRect() : {};

  return (
    <div className={styles.container}>
      <ExecutionHeader workflow={workflow} workflowExecution={workflowExecution} />
      <section aria-label="Executions" className={styles.executionResultContainer}>
        <ExecutionTaskLog workflowExecution={workflowExecution} />
        <div className={styles.executionDesignerContainer} ref={diagramRef}>
          <div className={styles.executionWorkflowActions}>
            <WorkflowActions workflow={workflow.data} />
            <WorkflowZoom
              //@ts-ignore
              workflowDagBoundingClientRect={workflowDagBoundingClientRect}
              workflowDagEngine={workflowDagEngine}
            />
          </div>
          <DiagramWidget
            allowLooseLinks={false}
            allowCanvasTranslation={true}
            allowCanvasZoom={true}
            className={styles.diagram}
            deleteKeys={[]}
            diagramEngine={workflowDagEngine.getDiagramEngine()}
            maxNumberPointsPerLink={0}
          />
          {!isDiagramLoading && (
            <div className={styles.diagramLoading}>
              <Loading withOverlay={false} />
            </div>
          )}
        </div>
      </section>
    </div>
  );
};

export default Main;
