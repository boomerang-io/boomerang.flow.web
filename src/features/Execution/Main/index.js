import React, { Component } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import { DiagramWidget } from "@projectstorm/react-diagrams";
import ExecutionHeader from "./ExecutionHeader";
import ExecutionTaskLog from "./ExecutionTaskLog";
import WorkflowActions from "./WorkflowActions";
import Loading from "Components/Loading";
import WorkflowZoom from "Components/WorkflowZoom";
import { REQUEST_STATUSES } from "Config/servicesConfig";
import DiagramApplication from "Utilities/DiagramApplication";
import { EXECUTION_STATUSES } from "Constants/workflowExecutionStatuses";
import styles from "./main.module.scss";

class Main extends Component {
  static propTypes = {
    dag: PropTypes.object.isRequired,
    workflow: PropTypes.object.isRequired,
    workflowExecution: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);
    this.diagramApp = new DiagramApplication({ dag: props.dag, modelIsLocked: true });
    this.state = {
      diagramBoundingClientRect: {}
    };
    this.diagramRef = React.createRef();
  }

  componentDidMount() {
    if (this.diagramRef.current) {
      this.setState({
        diagramBoundingClientRect: this.diagramRef.current.getBoundingClientRect()
      });
    }
    this.diagramApp.getDiagramEngine().zoomToFit();
  }

  render() {
    const { workflow, workflowExecution } = this.props;
    const { status, steps } = workflowExecution.data;

    const hasFinished = [EXECUTION_STATUSES.COMPLETED, EXECUTION_STATUSES.INVALID, EXECUTION_STATUSES.FAILURE].includes(
      status
    );

    const hasStarted = steps && steps.find(step => step.flowTaskStatus !== EXECUTION_STATUSES.NOT_STARTED);

    const loadDiagram =
      workflow.fetchingStatus === REQUEST_STATUSES.SUCCESS &&
      workflowExecution.status === REQUEST_STATUSES.SUCCESS &&
      (hasStarted || hasFinished);

    return (
      <div className={styles.container}>
        <ExecutionHeader workflow={workflow} workflowExecution={workflowExecution} />
        <main className={styles.executionResultContainer}>
          <ExecutionTaskLog workflowExecution={workflowExecution} />
          <div className={styles.executionDesignerContainer} ref={this.diagramRef}>
            <section className={styles.executionWorkflowActions}>
              <WorkflowActions workflow={workflow.data} />
              <WorkflowZoom
                diagramApp={this.diagramApp}
                diagramBoundingClientRect={this.state.diagramBoundingClientRect}
              />
            </section>
            <DiagramWidget
              allowLooseLinks={false}
              allowCanvasTranslation={true}
              allowCanvasZoom={true}
              className={styles.diagram}
              deleteKeys={[]}
              diagramEngine={this.diagramApp.getDiagramEngine()}
              maxNumberPointsPerLink={0}
            />
            {!loadDiagram && (
              <div className={styles.diagramLoading}>
                <Loading withOverlay={false} />
              </div>
            )}
          </div>
        </main>
      </div>
    );
  }
}

export default withRouter(Main);
