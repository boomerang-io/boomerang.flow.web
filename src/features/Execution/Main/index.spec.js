import React from "react";
import Main from "./index";

const props = {
  dag: {},
  workflow: {
    isFetching: false,
    data: {
      name: "Sparkle Flow with extra glitter and donuts on the side"
    },
    fetchingStatus: "success"
  },
  workflowExecution: {
    status: "success",
    data: {
      status: "inProgress",
      steps: []
    }
  }
};

describe("Main --- Snapshot", () => {
  it("Capturing Snapshot of Main", () => {
    const { baseElement } = rtlRouterRender(<Main {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
