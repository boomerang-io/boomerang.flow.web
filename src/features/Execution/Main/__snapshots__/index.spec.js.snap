// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main --- Snapshot Capturing Snapshot of Main 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="container"
    >
      <header
        class="container border"
      >
        <div
          class="container"
        >
          <section>
            <div
              class="subtitle"
            >
              <a
                href="/activity"
              >
                Activity
              </a>
              <p
                style="margin: 0px 0.5rem;"
              >
                /
              </p>
              <p>
                Sparkle Flow with extra glitter and donuts on the side
              </p>
            </div>
            <h1
              class="title"
            >
              Workflow run detail
            </h1>
          </section>
          <div
            class="content"
          >
            <div
              class="data"
            >
              <p
                class="dataTitle"
              >
                Team
              </p>
              <p
                class="dataValue"
              />
            </div>
            <div
              class="data"
            >
              <p
                class="dataTitle"
              >
                Initiated by
              </p>
              <span
                aria-hidden="false"
                aria-label="robot"
                role="img"
              >
                ðŸ¤–
              </span>
            </div>
            <div
              class="data"
            >
              <p
                class="dataTitle"
              >
                Trigger
              </p>
              <p
                class="dataValue"
              />
            </div>
            <div
              class="data"
            >
              <p
                class="dataTitle"
              >
                Team
              </p>
              <p
                class="dataValue"
              >
                2019-01-01 12:00 AM
              </p>
            </div>
          </div>
        </div>
      </header>
      <main
        class="executionResultContainer"
      >
        <aside
          class="container "
        >
          <section
            class="statusBlock inProgress"
          >
            <div
              class="duration"
            >
              <p
                class="title"
              >
                Duration
              </p>
              <time
                class="value"
              >
                --
              </time>
            </div>
            <div
              class="status"
            >
              <p
                class="title"
              >
                Status
              </p>
              <div
                class="statusData"
              >
                <svg
                  aria-label="inProgress"
                  class="statusIcon"
                  focusable="false"
                  height="32"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  style="will-change: transform;"
                  viewBox="0 0 32 32"
                  width="32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15 11h2v9h-2zm-2-9h6v2h-6z"
                  />
                  <path
                    d="M28 9l-1.42-1.41-2.25 2.25a10.94 10.94 0 1 0 1.18 1.65zM16 26a9 9 0 1 1 9-9 9 9 0 0 1-9 9z"
                  />
                </svg>
                <p
                  class="value"
                >
                  In Progress
                </p>
              </div>
            </div>
            <button
              class="collapseButton"
            >
              <svg
                aria-hidden="true"
                class="chevron"
                focusable="false"
                height="32"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 32 32"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 16L20 6l1.4 1.4-8.6 8.6 8.6 8.6L20 26z"
                />
              </svg>
            </button>
          </section>
          <section
            class="taskbar"
          >
            <p
              class="taskbarTitle"
            >
              Task log
            </p>
            <div
              class="bx--tooltip--definition bx--tooltip--a11y taskbarButton"
              data-testid="taskbar-button"
            >
              <button
                aria-describedby="definition-tooltip-2"
                class="bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition bx--tooltip--top bx--tooltip--align-center"
              >
                <svg
                  aria-hidden="true"
                  class="taskbarArrows"
                  focusable="false"
                  height="32"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 32 32"
                  width="32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M26.6 20.6L23 24.2V4h-2v20.2l-3.6-3.6L16 22l6 6 6-6zM10 4l-6 6 1.4 1.4L9 7.8V28h2V7.8l3.6 3.6L16 10z"
                  />
                </svg>
              </button>
              <div
                class="bx--assistive-text"
                id="definition-tooltip-2"
                role="tooltip"
              >
                Change sort direction (by start time)
              </div>
            </div>
          </section>
          <ul
            class="tasklog"
          />
        </aside>
        <div
          class="executionDesignerContainer"
        >
          <section
            class="executionWorkflowActions"
          >
            <div
              class="container"
            >
              <p
                class="messageText"
              >
                Read-only
              </p>
              <button
                class="editContainer"
              >
                <svg
                  aria-hidden="true"
                  class="editIcon"
                  focusable="false"
                  height="32"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 32 32"
                  width="32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M2 27h28v2H2zM25.41 9a2 2 0 0 0 0-2.83l-3.58-3.58a2 2 0 0 0-2.83 0l-15 15V24h6.41zm-5-5L24 7.59l-3 3L17.41 7zM6 22v-3.59l10-10L19.59 12l-10 10z"
                  />
                </svg>
                <p
                  class="editText"
                >
                  Edit Workflow
                </p>
              </button>
            </div>
            <div
              class="zoomIcons"
            >
              <button
                class="zoomButton"
              >
                <svg
                  aria-hidden="true"
                  class="zoomIcon"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 6h5v1H4z"
                  />
                  <path
                    d="M10.7 10c.8-1 1.3-2.2 1.3-3.5C12 3.5 9.5 1 6.5 1S1 3.5 1 6.5 3.5 12 6.5 12c1.3 0 2.5-.5 3.5-1.3l3.8 3.8.7-.7-3.8-3.8zm-4.2 1C4 11 2 9 2 6.5S4 2 6.5 2 11 4 11 6.5 9 11 6.5 11z"
                  />
                </svg>
              </button>
              <button
                class="zoomButton"
              >
                <svg
                  aria-hidden="true"
                  class="zoomIcon"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 6H7V4H6v2H4v1h2v2h1V7h2z"
                  />
                  <path
                    d="M10.7 10c.8-1 1.3-2.2 1.3-3.5C12 3.5 9.5 1 6.5 1S1 3.5 1 6.5 3.5 12 6.5 12c1.3 0 2.5-.5 3.5-1.3l3.8 3.8.7-.7-3.8-3.8zm-4.2 1C4 11 2 9 2 6.5S4 2 6.5 2 11 4 11 6.5 9 11 6.5 11z"
                  />
                </svg>
              </button>
            </div>
          </section>
          <div
            class="srd-diagram diagram "
          >
            <svg
              class="srd-link-layer "
              style="transform: translate(0px,0px) scale(NaN);"
            />
            <div
              class="srd-node-layer "
              style="transform: translate(0px,0px) scale(NaN);"
            />
          </div>
          <div
            class="diagramLoading"
          />
        </div>
      </main>
    </div>
  </div>
</body>
`;
